core/lib/zksync_core/src/consensus/storage.rs

```diff
@@ -0,0 +1,423 @@
+            .context("unknown block")?;
+        if payload != &want.encode() {
+            return Err(anyhow::anyhow!("unexpected payload").into());
+        }
+        Ok(())
+    }
+
+    /// Puts a block into this storage.
+    async fn put_block(&self, ctx: &ctx::Ctx, block: &validator::FinalBlock) -> ctx::Result<()> {
+        tracing::error!("SignedBlockStore::put_block()");
+        let storage = &mut storage(ctx, &self.pool).await.wrap("storage()")?;
+
+        // Currently main node is the only validator, so it should be the only one creating new
+        // blocks. To ensure that no gaps in the blocks are created we check here that we always
+        // insert a new head.
+        let head = *self.head.borrow();
+        let head = storage
+            .find_head_forward(ctx, head, self.operator_address)
+            .await
+            .wrap("find_head_forward()")?;
+        if block.header.number != head.header.number.next() {
+            return Err(anyhow::anyhow!(
+                "expected block with number {}, got {}",
+                head.header.number.next(),
+                block.header.number
+            )
+            .into());
+        }
+
+        storage
+            .put_block(ctx, block, self.operator_address)
+            .await
+            .wrap("put_block()")
+    }
+}
+
+#[async_trait::async_trait]
+impl BlockStore for SignedBlockStore {
+    async fn head_block(&self, ctx: &ctx::Ctx) -> ctx::Result<validator::FinalBlock> {
+        let storage = &mut storage(ctx, &self.pool).await.wrap("storage()")?;
+        let head = *self.head.borrow();
+        storage
+            .find_head_forward(ctx, head, self.operator_address)
+            .await
+            .wrap("find_head_forward()")
+    }
+
+    async fn first_block(&self, ctx: &ctx::Ctx) -> ctx::Result<validator::FinalBlock> {
+        let storage = &mut storage(ctx, &self.pool).await.wrap("storage()")?;
+        Ok(storage
+            .fetch_block(ctx, self.genesis, self.operator_address)
+            .await
+            .wrap("fetch_block()")?
+            .context("Genesis miniblock not present in Postgres")?)
+    }
+
+    async fn last_contiguous_block_number(
+        &self,
+        ctx: &ctx::Ctx,
+    ) -> ctx::Result<validator::BlockNumber> {
+        Ok(self.head_block(ctx).await?.header.number)
+    }
+
+    async fn block(
+        &self,
+        ctx: &ctx::Ctx,
+        number: validator::BlockNumber,
+    ) -> ctx::Result<Option<validator::FinalBlock>> {
+        let storage = &mut storage(ctx, &self.pool).await.wrap("storage()")?;
+        storage
+            .fetch_block(ctx, number, self.operator_address)
+            .await
+            .wrap("fetch_block()")
+    }
+
+    async fn missing_block_numbers(
+        &self,
+        _ctx: &ctx::Ctx,
+        _range: ops::Range<validator::BlockNumber>,
+    ) -> ctx::Result<Vec<validator::BlockNumber>> {
+        Ok(vec![]) // The storage never has missing blocks by construction
```

Comment:
Note that I've fixed this place for the EN block store. Although in practice we won't call `missing_block_numbers` with args that would lead to a non-empty response, this is non-local reasoning, so I've found it too shaky.

Reply:
done